client/state/data-layer/wpcom-http/utils.js

```diff
@@ -42,11 +43,45 @@ export const getHeaders = action => get( action, 'meta.dataLayer.headers', null
  */
 export const getProgress = action => get( action, 'meta.dataLayer.progress', null );
 
+export const makeParser = ( schema, schemaOptions = {}, transformer = identity ) => {
+	const options = Object.assign( { verbose: true }, schemaOptions );
+	const validator = schemaValidator( schema, options );
+
+	// filter out unwanted properties even though we may have let them slip past validation
+	// note: this property does not nest deeply into the data structure, that is, properties
+	// of a property that aren't in the schema could still come through since only the top
+	// level of properties are pruned
+	const filter = schemaValidator.filter( { ...schema, additionalProperties: false } );
+
+	// the actual parser
+	return responseData => {
+		if ( ! validator( responseData ) ) {
+			throw {
+				errorType: 'Failed to validate with JSON schema',
+				message: 'Failed to validate with JSON schema',
+				errors: validator.errors,
+			};
```

Comment:
@jsnajdr I like your idea of matching the `Error`, but at this point, why not just go ahead and throw an actual error?

```js
throw new TypeError(
	`Failed to validate with JSON schema: ${ validator.errors }`
);
```